src/openslide-vendor-dicom.c

```diff
@@ -698,6 +700,100 @@ static bool maybe_add_file(openslide_t *osr,
   }
 }
 
+struct property_iterate {
+  openslide_t *osr;
+  const char *prefix;
+};
+
+static bool add_properties_element(const DcmElement *element,
+                                   void *client);
+
+static bool add_properties_dataset(const DcmDataSet *dataset,
+                                   void *client) {
+  return dcm_dataset_foreach(dataset, add_properties_element, client);
+}
+
+static bool add_properties_element(const DcmElement *element,
+                                   void *client) {
+  struct property_iterate *iter = (struct property_iterate *) client;
+  DcmVR vr = dcm_element_get_vr(element);
+  uint32_t tag = dcm_element_get_tag(element);
+  const char *keyword = dcm_dict_keyword_from_tag(tag);
+  DcmVRClass klass = dcm_dict_vr_class(vr);
+
+  // ignore unknown tags
+  if (!keyword) {
+    return true;
+  }
+
+  if (klass == DCM_CLASS_SEQUENCE) {
+    DcmSequence *seq;
+    if (dcm_element_get_value_sequence(NULL, element, &seq)) {
+      char *new_prefix = g_strdup_printf("%s%s.", iter->prefix, keyword);
+      struct property_iterate new_iter = { iter->osr, new_prefix };
+      dcm_sequence_foreach(seq, add_properties_dataset, &new_iter);
+      g_free(new_prefix);
+    }
+  } else {
+    char *value = g_strdup("");
+    char *new_value;
+
+    switch (klass) {
+    case DCM_CLASS_STRING_MULTI:
+    case DCM_CLASS_STRING_SINGLE:
+      for (uint32_t i = 0; i < dcm_element_get_vm(element); i++) {
+        const char *str;
+        if (dcm_element_get_value_string(NULL, element, i, &str)) {
+          const char *format = i > 0 ? "%s, %s" : "%s%s";
+          new_value = g_strdup_printf(format, value, str);
+          g_free(value);
+          value = new_value;
+        }
+      }
+      break;
+
+    case DCM_CLASS_NUMERIC:
+      for (uint32_t i = 0; i < dcm_element_get_vm(element); i++) {
+        if (vr == DCM_VR_FL || vr == DCM_VR_FD) {
+          double d;
+          if (dcm_element_get_value_floatingpoint(NULL, element, i, &d)) {
+            const char *format = i > 0 ? "%s, %g" : "%s%g";
+            new_value = g_strdup_printf(format, value, d);
+            g_free(value);
+            value = new_value;
+          }
+        } else {
+          int64_t i64;
+          if (dcm_element_get_value_integer(NULL, element, i, &i64)) {
+            const char *format = i > 0 ? "%s, %"PRId64 : "%s%"PRId64;
+            new_value = g_strdup_printf(format, value, i64);
+            g_free(value);
+            value = new_value;
+          }
+        }
+      }
+      break;
+
+    case DCM_CLASS_BINARY:
+      new_value = g_strdup_printf("%s<binary value of %d bytes>", 
+                                  value,
+                                  dcm_element_get_length(element));
+      g_free(value);
+      value = new_value;
+      break;
+
+    default:
+      break;
+    }
+
+    g_hash_table_insert(iter->osr->properties,
```

Comment:
Is it physically possible for a file to have multiple properties with the same key?  If so, is libdicom guaranteed to give them to us in a consistent order?

Reply:
Keys are unique (phew) and always presented in numerical order (phew). It's just like TIFF.